### ISSUE TITLE
Audit challenges and make sure each challenge has a hint article

### ISSUE THREAD

**Issue by QuincyLarson** [freeCodeCamp/freeCodeCamp#16388]
Currently when you click the "get a hint" button, it opens a corresponding "hint article" on the forum with hints. We instead want to open a corresponding article on the Guide, since the Guide is much better suited toward presenting this type of information.

Currently, most of the hint articles lead to forum posts like this one: https://forum.freecodecamp.org/t/freecodecamp-algorithm-challenge-guide-roman-numeral-converter/16044

We've also ported these over to the guide, but they are varying states of repair: https://guide.freecodecamp.org/certificates/roman-numeral-converter/

@nsuchy has started the process of cleaning these up. We should finish it, then make sure every challenge on freeCodeCamp has at least a stub that we can point to for it.

This will be a big undertaking. I estimate it will be perhaps 20 - 30 hours worth of coding. These improvements could come through many pull requests.

In short, we need to:
- [ ] Clean up the existing Guide articles for each challenge
- [ ] Audit the challenges to make a list of challenges that lack a corresponding article
- [ ] Create stubs for challenges that lack an article
- [ ] Update the hint button on staging to point to the Guide rather than the forum

**Comment by iabrenne** [comment]
I would like to give this one a stab, if that's OK.

**Comment by QuincyLarson** [comment]
@iabrenne OK - awesome! Which part of this do you want to tackle first?

**Comment by iabrenne** [comment]
@QuincyLarson - I can start with with updating the links - the last item on the list:

 - [ ] Update the hint button on staging to point to the Guide rather than the forum

**Comment by iabrenne** [comment]
@QuincyLarson  I do not see "Get a Hint" button on the beta side, only on the production site.

![image](https://user-images.githubusercontent.com/24214306/34474805-d64fef18-ef49-11e7-89ce-068d872beb1b.png)

![image](https://user-images.githubusercontent.com/24214306/34474835-714f169c-ef4a-11e7-975c-c0e2d08c6ee8.png)

Beta site has the "Get Help" button, which opens up this modal : 

![image](https://user-images.githubusercontent.com/24214306/34474847-9e886032-ef4a-11e7-8058-44a5e3647651.png)

It does not seem to link to any articles, on the forum or otherwise.

Would like to clarify if we are we looking to update live site, beta, or both?  

Thank you!

**Comment by iabrenne** [comment]
@QuincyLarson  Just wanted to provide a quick update. I am familiarizing myself with the codebase. It's quite a lot to digest.

My plan so far is to leave the Tool Panel on the beta site with all current buttons intact, and add another button with "See Guide" or something to that effect. Once user clicks on the "See Guide" button, it'll take them to the corresponding challenge article in the guide. ( Not sure about "Get Hint" - there is already a "hints" array in the schema for Challenges and it seems like it may have been meant for something else so I don't feel right naming new button "Get Hint", plus per my post above I'm a little confused with "Get Hint" button present in prod site but not beta. )

I'm still digging around but so far it seems like the schema for the challenges is in the /common/models/challenge.json and defaults are seeded in the seed/challenges/ .json files so I'm thinking in the very least those two would need to be updated. If I'm missing something or have some guidance please let me know.

Thank you!

**Comment by QuincyLarson** [comment]
@iabrenne Thanks for your extreme patience. I forgot to respond to you here. It looks like you're making steady progress on this, and that Berkeley is helping you QA your pull request. Please let me know if I can still be of any help :)

**Comment by iabrenne** [comment]
Thank you, @QuincyLarson - I will definitely let you know if I have questions or need help  :)

**Comment by QuincyLarson** [comment]
@iabrenne How is this process coming along? One thing I didn't mention is that it would probably be easier to have all the articles follow a convention (having the Guide articles use the same name as challenges) so that we don't have to specify a guide article URL for each challenge, and the guide article URL can be inferred from the current challenge's name.

**Comment by iabrenne** [comment]
@QuincyLarson I did not think of that approach. Instead I've been going through each article in the guide in alphabetical order ( currently on "f" ) and adding guideUrl property into that challenge object. 

Once I was done adding all the available urls, I was planning on making a list of challenges lacking a guide. I have a mongo script looking for challenges missing guideUrl field :) It's not perfect as it does not account for challenges that should NOT have a guide article, like intro challenges, etc but was a good starting point.

Wondering if we should scrap the PR in start over. Definitely see what you are saying - my approach is not very dynamic and it would take a long time to add guideUrl field to all the challenges. There are currently 1810 challenges total ( still a big number, even if we exclude certain challenge types intros, etc ), and so far I've added 84 urls ( 20 in the existing PR and rest on my local branch ) .

On the other hand, this approach helps find challenges missing articles and allows us to control which articles to add challenges to .

The more I think about it, the more sense it makes to start over. Instead of starting from the bottom ( adding link to guide ) , perhaps I should go in order you suggested originally, starting with "Clean up the existing Guide articles for each challenge". Which would basically involve making sure that guide article name matches challenge name minus the dashes. 

Let me know what you think! Thank you.

**Comment by iabrenne** [comment]
@QuincyLarson In the meantime, my plan of action for item 1 ( Clean up the existing Guide articles for each challenge ). 

1. Get list of all article names under https://guide.freecodecamp.org/certificates , remove hyphens and sort alphabetically

2. Get a list of all challenge titles from freecodecamp. 

3. Merge the two lists and get list that exists in 1 but not 2. I expect those will be guide articles that have a name that doesn't match their corresponding challenge. Or perhaps somehow we have some extras ( guide for a non existing challenge...weird, but who knows...)

4. PRs to update the names/URLs of the identified guide articles so that it matches to the corresponding challenge.

Please let me know if I am missing anything. This is just for step 1 above, assuming we are going to scrap PR https://github.com/freeCodeCamp/freeCodeCamp/pull/16442 and start over.

Thank you.

**Comment by iabrenne** [comment]
@QuincyLarson 

1. Got a list of all guide articles from https://guide.freecodecamp.org/certificates   ( count : 261 )
2. Got a list of all challenge titles from challenge collection in mongo ( count : 1887 ) . ( # will go down once I figure out which challengeTypes  to exclude e.g. intros, etc...)
3. Compared two files with diff -y

Out of 261 guide articles, 
       195 have name/url that matches perfectly with the corresponding challenge name ( after adjusting for hyphens and capitalization )
       66 do not have a name that matches perfectly with the corresponding challenge name 

I'm thinking my next step should be going through the 66 guide articles and renaming them so that they match with their corresponding challenge, then submitting PR for that update.


[guide_challenge_name_match(195).txt](https://github.com/freeCodeCamp/freeCodeCamp/files/1633964/guide_challenge_name_match.195.txt)
[in_guide_not_challenge(66).txt](https://github.com/freeCodeCamp/freeCodeCamp/files/1633965/in_guide_not_challenge.66.txt)

Few questions :  

Am I on the right track?

Since Guide is a seperate repo - would I be able to submit a PR from this issue?

Appears there are >3K PRs in the guide repo. PR for this issue would take a while to get merged, no?

Again, this is all addressing item 1 (  Clean up the existing Guide articles for each challenge ) in the original issue description.

**Comment by QuincyLarson** [comment]
@iabrenne Thanks for your patience - it sounds like you've gotten a ton done in the meantime!

> 195 have name/url that matches perfectly with the corresponding challenge name ( after adjusting for hyphens and capitalization )

Great! I figured this would be the case, since all of these articles were exported from the https://forum.freecodecamp.org/c/guide forum category.

> 66 do not have a name that matches perfectly with the corresponding challenge name

Some of the challenge names may have changed slightly. We can update the titles of the guide articles accordingly. The guide articles should track the challenges (and not the other way around) since it's much simpler to change a guide article than it is to change a challenge.

> I'm thinking my next step should be going through the 66 guide articles and renaming them so that they match with their corresponding challenge, then submitting PR for that update.

Great - that sounds like a good next step.

> Am I on the right track?

Most definitely.

> Since Guide is a seperate repo - would I be able to submit a PR from this issue?

You would submit the renaming PR to the guide repo. We will also need to update the freeCodeCamp/freeCodeCamp repo so that the "get a hint" button sends campers to the corresponding Guide article rather than the forum article. 

> Appears there are >3K PRs in the guide repo. PR for this issue would take a while to get merged, no?

This is correct, but I doubt many - if any - of these PRs will be impacted by renaming some of these guide articles. We are steadily QA'ing and merging these PRs, but for your purposes, you can ignore them.

> Again, this is all addressing item 1 ( Clean up the existing Guide articles for each challenge ) in the original issue description.

Yes.

I've opened an issue to, among other things, update beta to link to guide.freecodecamp.org articles instead of forum articles.

**Comment by iabrenne** [comment]
@QuincyLarson  Thank you so much for your response and guidance. I will proceed with submitting PRs against the Guide repo to get the names of the 66 guide articles changed to match the challenge names.

Few questions in the meantime.

Do I understand it correctly that we want to have challenges linked to guide articles on both beta and production? ( with updates to beta being tracked in the [16517](https://github.com/freeCodeCamp/freeCodeCamp/issues/16517) , and updates to production being tracked in step 4 in the description of this issue - "Update the hint button on staging to point to the Guide rather than the forum"? )

If so, would master branch be the one to modify for production? "Staging" is mentioned in step 4 of the description but seems like staging branch is for beta.

Also, my earlier [PR](https://github.com/freeCodeCamp/freeCodeCamp/pull/16442) was merged ( oops ) which I'm guessing would need to be backed out, since it added (yet) another button into beta challenges. Button is labeled "See Guide" and so far only shows up for the 20 challenges that I've added url property for ( that was before we decided to go w/ the dynamic approach of deriving url from a challenge name. 

![image](https://user-images.githubusercontent.com/24214306/35076071-13527c68-fbbc-11e7-903e-9f25c8f427b6.png)

I'm thinking removing code introduced in that PR could be done as part of (https://github.com/freeCodeCamp/freeCodeCamp/issues/16517) you created for beta updates. Let me know what you think.

Also, when we get to step 3 ( creating stubs for challenges lacking an article ), looking at the challengeTypes:

export const html = 0;
export const js = 1;
export const oldVideo = 2;
export const simpleProject = 2;
export const zipline = 3;
export const frontEndProject = 3;
export const basejump = 4;
export const backEndProject = 4;
export const bonfire = 5;
export const video = 6;
export const step = 7;
export const quiz = 8;

I'm thinking we want 0,1,2 3 and 4 challengeTypes linked to guide - would you agree?  I don't want to add links to challenges that aren't really challenges ( e.g. intros, etc ) but am not 100% sure how to identify those. Thank you.

**Comment by QuincyLarson** [comment]
@iabrenne Sure - let me answer your questions one-by-one:

> Do I understand it correctly that we want to have challenges linked to guide articles on both beta and production? ( with updates to beta being tracked in the 16517 , and updates to production being tracked in step 4 in the description of this issue - "Update the hint button on staging to point to the Guide rather than the forum"? )
> If so, would master branch be the one to modify for production? "Staging" is mentioned in step 4 of the description but seems like staging branch is for beta.

Don't worry about production - we have only been making changes to production when absolutely necessary. The staging branch will soon become production.

> I'm thinking removing code introduced in that PR could be done as part of (#16517) you created for beta updates. Let me know what you think.

Yes - no need to remove it. @Bouncey is going to reconfigure the buttons anyway.

> I'm thinking we want 0,1,2 3 and 4 challengeTypes linked to guide - would you agree? I don't want to add links to challenges that aren't really challenges ( e.g. intros, etc ) but am not 100% sure how to identify those. Thank you.

That's correct. On staging, there are no more type 2 or 6, and type 5 will probably get converted to type 1 soon. This entire type system may go away with a near-future refactor. But I digress :)

**Comment by iabrenne** [comment]
@QuincyLarson  - thank you. Sounds like item 4 in the description of this issue ( actual updating of the FCC code to link challenges to guide articles ) is being handled by @Bouncey in #16517.  I will focus on cleaning up the guide repo - renaming articles to match challenge titles and creating stubs for the missing challenges. Thanks again.

**Comment by QuincyLarson** [comment]
@iabrenne I just merged your PR. Great work so far! Yes - I think going ahead and creating the stubs to match the titles is all that remains to be able to close this issue :)

**Comment by QuincyLarson** [comment]
@iabrenne Thanks for the update. I'm sad to hear you caught the flue. Please get plenty of rest and focus on recovering. Feel better soon!

**Comment by QuincyLarson** [comment]
@iabrenne Just a quick update - we are still waiting for the new version of Gatsby.js, which will help us fix the problem we're having with too many article in Guide. As soon as we do, we'll resume working on all this. I'm closing this issue for now. Thanks for all your help.

By the way, we just launched our new curriculum and learning platform: https://learn.freecodecamp.org. We would welcome further contributions from you. Thanks again for helping improve the freeCodeCamp open source community!
